addons/web_tour/static/src/js/tip.js

```diff
@@ -113,16 +124,100 @@ var Tip = Widget.extend({
+                    this.$furtherIdealLocation = this._get_ideal_location(this.$ideal_location);
+                    this.$el.appendTo(this.$ideal_location);
+                } else {
+                    this.$ideal_location = this._get_ideal_location($anchor);
+                    this.$furtherIdealLocation = this._get_ideal_location(this.$ideal_location);
+                    this.$actualAnchor = this.$ideal_location;
+                    this.$el.appendTo(this.$furtherIdealLocation);
+                }
             }
             this._bind_anchor_events();
         }
         this._reposition();
     },
-    _get_ideal_location: function () {
-        var $location = this.$anchor;
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * This function figure out which direction the tip should take
+     * and if he's at the bottom or the top of the targeted element
+     * or if it's an indicator to scroll.
+     *
+     * If the direction or the anchor change, then we adapth the frontend
+     * position of the tip.
+     */
+    _update_position: function () {
+        this.moveTip = false;
+        let halfHeight = 0;
+        if (this.initialPosition === 'right' || this.initialPosition === 'left') {
+            halfHeight = this.$anchor.innerHeight() / 2;
+        }
+        let position = this.info.position;
+        /* Check if the viewport state change to know if we need to move the anchor of the tip.
+        * up : the target element is above the current viewport
+        * down : the target element is below the current viewport
+        * in : the target element is in the current viewport
+        */
+        let viewPortState = "in";
+        const paddingTop = this.$ideal_location.css('padding-top') ? parseInt(this.$ideal_location.css('padding-top')) : 0;
+        if (this.$anchor.offset().top + this.$anchor.innerHeight() - paddingTop - halfHeight < window.pageYOffset) {
+            viewPortState = 'up';
+        } else if (this.$anchor.offset().top + halfHeight > window.pageYOffset + window.innerHeight) {
+            viewPortState = 'down';
+        }
+        if (this.viewPortState !== viewPortState) { // If the state change, allow to move the tip at the right place
+            this.viewPortState = viewPortState;
+            this.moveTip = true;
+        }
+
+        // Adjust the placement of the tip regarding its anchor depending
+        // if we came from the bottom or the top.
+        if (this.$anchor.offset().top < window.pageYOffset + paddingTop + this.$el.innerHeight()) {
+            position = halfHeight ? this.initialPosition : "bottom";
+        } else if (this.$anchor.offset().top > window.pageYOffset + window.innerHeight - (this.$anchor.innerHeight() + this.$el.innerHeight())) {
+            position = halfHeight ? this.initialPosition : "top";
+        }
```

Comment:
Not sure about the utility of this code, won't this be handled automatically with the lines below? Checking the `viewPortState`?

Reply:
It's handled when the anchor is not in the viewport.
I added these lines to move the tip when the anchor is still in the viewport but not entirely and so, the tip is out of the viewport.
By example, my position is 'top', but I only see bottom half of the anchor object in my window. These line move the tip at the 'bottom'.

Reply:
Still don't understand. If the anchor is below the viewport, you have a tip with position: top pointing at the scroll container. Then the user scroll, and the anchor is half visible... the position should still be top without any other computation, so why would these lines be needed?

Reply:
I guess I'll show you in person and we will figure out what we do about there lines after that.